In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv


In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn import tree
from sklearn.model_selection import train_test_split

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [3]:
df=pd.read_csv('/kaggle/input/titanic/train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
df=df.drop('Name',axis='columns')
df=df.drop('SibSp',axis='columns')
df=df.drop('Parch',axis='columns')
df=df.drop('Ticket',axis='columns')
df=df.drop('Cabin',axis='columns')
df=df.drop('Embarked',axis='columns')
df

PassengerId  Survived  Pclass     Sex   Age     Fare
0              1         0       3    male  22.0   7.2500
1              2         1       1  female  38.0  71.2833
2              3         1       3  female  26.0   7.9250
3              4         1       1  female  35.0  53.1000
4              5         0       3    male  35.0   8.0500
..           ...       ...     ...     ...   ...      ...
886          887         0       2    male  27.0  13.0000
887          888         1       1  female  19.0  30.0000
888          889         0       3  female   NaN  23.4500
889          890         1       1    male  26.0  30.0000
890          891         0       3    male  32.0   7.7500

[891 rows x 6 columns]

In [5]:
df.isnull().sum()



PassengerId      0
Survived         0
Pclass           0
Sex              0
Age            177
Fare             0
dtype: int64

In [6]:
a=df.Age.median()
df.Age=df.Age.fillna(a)
df

PassengerId  Survived  Pclass     Sex   Age     Fare
0              1         0       3    male  22.0   7.2500
1              2         1       1  female  38.0  71.2833
2              3         1       3  female  26.0   7.9250
3              4         1       1  female  35.0  53.1000
4              5         0       3    male  35.0   8.0500
..           ...       ...     ...     ...   ...      ...
886          887         0       2    male  27.0  13.0000
887          888         1       1  female  19.0  30.0000
888          889         0       3  female  28.0  23.4500
889          890         1       1    male  26.0  30.0000
890          891         0       3    male  32.0   7.7500

[891 rows x 6 columns]

In [7]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [8]:
df['gender']=le.fit_transform(df.Sex)
df


PassengerId  Survived  Pclass     Sex   Age     Fare  gender
0              1         0       3    male  22.0   7.2500       1
1              2         1       1  female  38.0  71.2833       0
2              3         1       3  female  26.0   7.9250       0
3              4         1       1  female  35.0  53.1000       0
4              5         0       3    male  35.0   8.0500       1
..           ...       ...     ...     ...   ...      ...     ...
886          887         0       2    male  27.0  13.0000       1
887          888         1       1  female  19.0  30.0000       0
888          889         0       3  female  28.0  23.4500       0
889          890         1       1    male  26.0  30.0000       1
890          891         0       3    male  32.0   7.7500       1

[891 rows x 7 columns]

In [9]:
df=df.drop('Sex',axis='columns')
df

PassengerId  Survived  Pclass   Age     Fare  gender
0              1         0       3  22.0   7.2500       1
1              2         1       1  38.0  71.2833       0
2              3         1       3  26.0   7.9250       0
3              4         1       1  35.0  53.1000       0
4              5         0       3  35.0   8.0500       1
..           ...       ...     ...   ...      ...     ...
886          887         0       2  27.0  13.0000       1
887          888         1       1  19.0  30.0000       0
888          889         0       3  28.0  23.4500       0
889          890         1       1  26.0  30.0000       1
890          891         0       3  32.0   7.7500       1

[891 rows x 6 columns]

In [10]:
a=df[df['Survived']==1]
a

PassengerId  Survived  Pclass   Age     Fare  gender
1              2         1       1  38.0  71.2833       0
2              3         1       3  26.0   7.9250       0
3              4         1       1  35.0  53.1000       0
8              9         1       3  27.0  11.1333       0
9             10         1       2  14.0  30.0708       0
..           ...       ...     ...   ...      ...     ...
875          876         1       3  15.0   7.2250       0
879          880         1       1  56.0  83.1583       0
880          881         1       2  25.0  26.0000       0
887          888         1       1  19.0  30.0000       0
889          890         1       1  26.0  30.0000       1

[342 rows x 6 columns]

In [11]:
b=df[df['Survived']==0]
b

PassengerId  Survived  Pclass   Age     Fare  gender
0              1         0       3  22.0   7.2500       1
4              5         0       3  35.0   8.0500       1
5              6         0       3  28.0   8.4583       1
6              7         0       1  54.0  51.8625       1
7              8         0       3   2.0  21.0750       1
..           ...       ...     ...   ...      ...     ...
884          885         0       3  25.0   7.0500       1
885          886         0       3  39.0  29.1250       0
886          887         0       2  27.0  13.0000       1
888          889         0       3  28.0  23.4500       0
890          891         0       3  32.0   7.7500       1

[549 rows x 6 columns]

In [12]:
a.describe()

PassengerId  Survived      Pclass         Age        Fare      gender
count   342.000000     342.0  342.000000  342.000000  342.000000  342.000000
mean    444.368421       1.0    1.950292   28.291433   48.395408    0.318713
std     252.358840       0.0    0.863321   13.764425   66.596998    0.466660
min       2.000000       1.0    1.000000    0.420000    0.000000    0.000000
25%     250.750000       1.0    1.000000   21.000000   12.475000    0.000000
50%     439.500000       1.0    2.000000   28.000000   26.000000    0.000000
75%     651.500000       1.0    3.000000   35.000000   57.000000    1.000000
max     890.000000       1.0    3.000000   80.000000  512.329200    1.000000

In [13]:
b.describe()

PassengerId  Survived      Pclass         Age        Fare      gender
count   549.000000     549.0  549.000000  549.000000  549.000000  549.000000
mean    447.016393       0.0    2.531876   30.028233   22.117887    0.852459
std     260.640469       0.0    0.735805   12.499986   31.388207    0.354968
min       1.000000       0.0    1.000000    1.000000    0.000000    0.000000
25%     211.000000       0.0    2.000000   23.000000    7.854200    1.000000
50%     455.000000       0.0    3.000000   28.000000   10.500000    1.000000
75%     675.000000       0.0    3.000000   35.000000   26.000000    1.000000
max     891.000000       0.0    3.000000   74.000000  263.000000    1.000000

In [14]:
df.groupby('gender').mean()

PassengerId  Survived    Pclass        Age       Fare
gender                                                       
0        431.028662  0.742038  2.159236  27.929936  44.479818
1        454.147314  0.188908  2.389948  30.140676  25.523893

from here we can say the 74.2% women survived whereas only 18.8 % men survived 

In [15]:
X_train=df.drop('Survived',axis='columns')
y_train=df["Survived"]
X_train

PassengerId  Pclass   Age     Fare  gender
0              1       3  22.0   7.2500       1
1              2       1  38.0  71.2833       0
2              3       3  26.0   7.9250       0
3              4       1  35.0  53.1000       0
4              5       3  35.0   8.0500       1
..           ...     ...   ...      ...     ...
886          887       2  27.0  13.0000       1
887          888       1  19.0  30.0000       0
888          889       3  28.0  23.4500       0
889          890       1  26.0  30.0000       1
890          891       3  32.0   7.7500       1

[891 rows x 5 columns]

In [16]:
test=pd.read_csv('/kaggle/input/titanic/test.csv')
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [17]:
test=test.drop('Name',axis='columns')
test=test.drop('SibSp',axis='columns')
test=test.drop('Parch',axis='columns')
test=test.drop('Ticket',axis='columns')
test=test.drop('Cabin',axis='columns')
test=test.drop('Embarked',axis='columns')
test

PassengerId  Pclass     Sex   Age      Fare
0            892       3    male  34.5    7.8292
1            893       3  female  47.0    7.0000
2            894       2    male  62.0    9.6875
3            895       3    male  27.0    8.6625
4            896       3  female  22.0   12.2875
..           ...     ...     ...   ...       ...
413         1305       3    male   NaN    8.0500
414         1306       1  female  39.0  108.9000
415         1307       3    male  38.5    7.2500
416         1308       3    male   NaN    8.0500
417         1309       3    male   NaN   22.3583

[418 rows x 5 columns]

In [18]:
test.isnull().sum()

PassengerId     0
Pclass          0
Sex             0
Age            86
Fare            1
dtype: int64

In [19]:
m=test.Age.median()
z=test.Fare.median()


In [20]:
test.Age=test.Age.fillna(m)
test.Fare=test.Fare.fillna(z)
test

PassengerId  Pclass     Sex   Age      Fare
0            892       3    male  34.5    7.8292
1            893       3  female  47.0    7.0000
2            894       2    male  62.0    9.6875
3            895       3    male  27.0    8.6625
4            896       3  female  22.0   12.2875
..           ...     ...     ...   ...       ...
413         1305       3    male  27.0    8.0500
414         1306       1  female  39.0  108.9000
415         1307       3    male  38.5    7.2500
416         1308       3    male  27.0    8.0500
417         1309       3    male  27.0   22.3583

[418 rows x 5 columns]

In [21]:
test['gender']=le.fit_transform(test.Sex)
test=test.drop('Sex',axis='columns')
test


PassengerId  Pclass   Age      Fare  gender
0            892       3  34.5    7.8292       1
1            893       3  47.0    7.0000       0
2            894       2  62.0    9.6875       1
3            895       3  27.0    8.6625       1
4            896       3  22.0   12.2875       0
..           ...     ...   ...       ...     ...
413         1305       3  27.0    8.0500       1
414         1306       1  39.0  108.9000       0
415         1307       3  38.5    7.2500       1
416         1308       3  27.0    8.0500       1
417         1309       3  27.0   22.3583       1

[418 rows x 5 columns]

In [22]:
testpassenderid=test['PassengerId']
testpassenderid

0       892
1       893
2       894
3       895
4       896
       ... 
413    1305
414    1306
415    1307
416    1308
417    1309
Name: PassengerId, Length: 418, dtype: int64

In [23]:
model=tree.DecisionTreeClassifier()
model.fit(X_train,y_train)
ans=model.predict(test)
ans

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [24]:
y_test=pd.read_csv('/kaggle/input/titanic/gender_submission.csv')
y_test

PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [25]:
import pandas as pd
y_test1=y_test.drop('PassengerId',axis='columns')
y_test1


Survived
0           0
1           1
2           0
3           0
4           1
..        ...
413         0
414         1
415         0
416         0
417         0

[418 rows x 1 columns]

In [26]:
y1=y_test['PassengerId']
y1

0       892
1       893
2       894
3       895
4       896
       ... 
413    1305
414    1306
415    1307
416    1308
417    1309
Name: PassengerId, Length: 418, dtype: int64

In [27]:
ans1=pd.DataFrame({'Survived':ans})
ans1

Survived
0           0
1           0
2           0
3           0
4           1
..        ...
413         0
414         1
415         0
416         0
417         0

[418 rows x 1 columns]

In [28]:
final=pd.concat([y1,ans1],axis='columns')
final

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [29]:
final.set_index('PassengerId',inplace=True)

In [30]:
final.to_csv("titanicsubmission1.csv")

In [31]:
ans1

Survived
0           0
1           0
2           0
3           0
4           1
..        ...
413         0
414         1
415         0
416         0
417         0

[418 rows x 1 columns]

In [32]:
y_test1

Survived
0           0
1           1
2           0
3           0
4           1
..        ...
413         0
414         1
415         0
416         0
417         0

[418 rows x 1 columns]

In [33]:
ylist=y_test1.values.tolist()

In [34]:
alist=ans1.values.tolist()

In [35]:
model.score(test,ylist)

0.8229665071770335

In [36]:
from sklearn.ensemble import RandomForestClassifier
model1=RandomForestClassifier(n_estimators=25)

In [37]:
model1.fit(X_train,y_train)

RandomForestClassifier(n_estimators=25)

In [38]:
ans2=model.predict(test)
ans2

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [39]:
ans2=pd.DataFrame({'Survived':ans2})
ans2

Survived
0           0
1           0
2           0
3           0
4           1
..        ...
413         0
414         1
415         0
416         0
417         0

[418 rows x 1 columns]

In [40]:
final2=pd.concat([y1,ans2],axis='columns')
final2

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [41]:
final2.set_index('PassengerId',inplace=True)
final2

Survived
PassengerId          
892                 0
893                 0
894                 0
895                 0
896                 1
...               ...
1305                0
1306                1
1307                0
1308                0
1309                0

[418 rows x 1 columns]

In [42]:
final2.to_csv("submission.csv")

In [43]:
model.score(test,ylist)

0.8229665071770335